src/glue.rs

```diff
@@ -15,24 +15,24 @@ impl Glue {
         Self { storage }
     }
 
-    pub fn execute(&mut self, sql: &str) -> Result<Payload> {
-        let parsed = parse(sql)?;
-        let statement = translate(&parsed[0])?;
-
-        let storage = self.storage.take().unwrap();
-
-        match block_on(execute(storage, &statement)) {
-            Ok((storage, payload)) => {
-                self.storage = Some(storage);
-
-                Ok(payload)
-            }
-            Err((storage, error)) => {
-                self.storage = Some(storage);
-
-                Err(error)
+    pub fn execute(&mut self, sql: &str) -> (Vec<Result<Payload>>, Option<Result<Payload>>) {
+        let mut results: Vec<Result<Payload>> = Vec::new();
+        let parsed = parse(sql);
+        for a_sql in parsed.unwrap().iter() {
+            let statement = translate(a_sql).unwrap();
+            let storage = self.storage.take().unwrap();
+            match block_on(execute(storage, &statement)) {
+                Ok((storage, payload)) => {
+                    self.storage = Some(storage);
+                    results.push(Ok(payload));
+                }
```

Comment:
You can declare `let mut storage = self.storage.take().unwrap();` outside of `for` loop and re-use it.
Then, you don't need to repeat `self.storage = Some(storage)` -> `let storage = self.storage.take().unwrap();`

Reply:
I'm struggling with this but have two issues.
1. `exeucte` does not take `storage` as a reference which means we can not `borrow` (`move` fails cuz of iteration).
2. It seems like `self.storage` is mutated at each iteration. I wonder it is okay not to use `self.storage.take().unwrap()` at each iteration as well.

I hope you to give me some advice !

Reply:
In this case, you don't need to concern about borrowing.
`execute` takes the ownership of `storage`, not mutable borrowing.
This is why the result of `execute` includes `Storage` instance rather than simply returning `Payload`.
It is for making execution layer codes to be pure functional without using `mut`.

`self.storage = Some(storage)` assignment is required only at the end of execution.

for example,
```rust
let parsed = parse(sql).unwrap();
let mut storage = self.storage.take().unwrap();
let mut payload: Option<Payload> = None;

for statement in parsed.iter() {
  let statement = translate(statement).unwrap();
  match block_on(execute(storage, &statement)) {
    Ok((s, p)) => {
      storage = s;
      payload = Some(p);
    }
    Err((s, e)) => {
      self.storage = Some(storage);
      return Err(e);
    }
  }
}

self.storage = Some(storage);
return Ok(payload.unwrap());
```
Something this kind of codes are which you can think about to use.

fyi.
This is just for extra info, it's ok to use `mut` at this time.
If you want something more fun stuff, then you can try iterator `try_fold` to solve this issue.
Then you don't even need to use `mut`.
`parsed.iter().try_fold(storage, ...)`
In the execution layer codes, `mut` is not allowed so `try_fold` is being quite commonly used.
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold